src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,198 @@
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "../storage/table.hpp"
+#include "abstract_operator.hpp"
+#include "all_type_variant.hpp"
+#include "resolve_type.hpp"
+#include "storage/dictionary_segment.hpp"
+#include "storage/reference_segment.hpp"
+#include "table_scan.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator> in, ColumnID column_id, const ScanType scan_type,
+                     const AllTypeVariant search_value)
+    : AbstractOperator(in, nullptr), _column_id(column_id), _scan_type(scan_type), _search_value(search_value) {}
+
+ColumnID TableScan::column_id() const { return _column_id; }
+
+ScanType TableScan::scan_type() const { return _scan_type; }
+
+const AllTypeVariant& TableScan::search_value() const { return _search_value; }
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+
+  const auto in_table = _input_table_left();
+  auto out_table = std::make_shared<Table>();
+  auto posList = PosList();
+  
+  auto comparator = [](ScanType scanType, auto left, auto right) {
+    switch (scanType) {
+      case ScanType::OpEquals: {
+        return left == right;
+        break;
+      }
+      case ScanType::OpNotEquals: {
+        return left != right;
+        break;
+      }
+      case ScanType::OpLessThan: {
+        return left < right;
+        break;
+      }
+      case ScanType::OpLessThanEquals: {
+        return left <= right;
+        break;
+      }
+      case ScanType::OpGreaterThan: {
+        return left > right;
+        break;
+      }
+      case ScanType::OpGreaterThanEquals: {
+        return left >= right;
+        break;
+      }
+    }
+    return left > right;
+  };
+
+  for (ChunkID i = ChunkID(0); i < in_table->chunk_count(); i++) {
+    const auto& chunk = in_table->get_chunk(i);
+    const auto& segment = chunk.get_segment(_column_id);
+
+    auto data_type = in_table->column_type(_column_id);
+    
+    resolve_data_type(data_type, [&] (auto type) {
+      using Type = typename decltype(type)::type;
+
+      const auto typed_value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      if(typed_value_segment != nullptr) {
+
+          auto values = typed_value_segment->values();
+          for (size_t index = 0; index < values.size(); index++){
+            auto value = values.at(index);
```

Comment:
at() is too expensive to be done in the hot loop. You could assert the size of the vector outside of the loop.